File: Element/JumpGame.cpp
// Jump Game: O(N) complexity

#include <bits/stdc++.h>
using namespace std;

class Solution{
	public:
	static bool isReachable(vector<int> &arr){
		int n = arr.size();
		if(n <= 1)
			return true;
		int maxReach = 0;
		int step = 1;
		for(int i = 0; i < n; ++i){
			step--;
			if(i+arr[i] > maxReach){
				maxReach = i+arr[i];
				step = arr[i];
			}
			if(step == 0 && i < n-1)
				return false;
		}
		return true;
	}
};
int main(){
	vector<int> test1 = {2,3,1,1,4};
	vector<int> test2 = {3,2,1,0,4};
	printf("%d\n",Solution::isReachable(test1));
	printf("%d",Solution::isReachable(test2));
}

File: Element/JumpGame.py
class Solution(object):
    def __init__(self,arr):
        self.arr = arr
    def isReachable(self,arr)->bool:
        self.n = len(self.arr)
        if(self.n<=1):
            return True
        self.maxReach = 0
        self.step = 1
        for i in range(self.n):
            self.step -= 1
            if(i+self.arr[i] > self.maxReach):
                self.maxReach = i+self.arr[i]
                self.step = self.arr[i]
            if(self.step == 0 and i < self.n-1):
                return False
        return True
    

if __name__=='__main__':
    test1 = [2,3,1,1,4]
    test2 = [3,2,1,0,4]
    s1 = Solution(test1)
    s2 = Solution(test2)
    print(s1.isReachable(test1)) # True
    print(s2.isReachable(test2)) # False
